src/heap_config.cpp

```diff
@@ -0,0 +1,132 @@
+    }
+
+    return default_value;
+}
+} // namespace detail
+
+heap_config::heap_config(bool never_free, std::size_t num_reserve_segments, std::size_t tiny_limit,
+    std::size_t small_limit, std::size_t large_limit, std::size_t tiny_segment_size,
+    std::size_t small_segment_size, std::size_t large_segment_size, std::size_t num_huge_alloc_warn_threshold)
+: m_never_free{never_free}
+, m_num_reserve_segments{num_reserve_segments}
+, m_tiny_limit{detail::round_to_pow_of_2(tiny_limit)}
+, m_small_limit{detail::round_to_pow_of_2(small_limit)}
+, m_large_limit{detail::round_to_pow_of_2(large_limit)}
+, m_tiny_segment_size{detail::round_to_pow_of_2(tiny_segment_size)}
+, m_small_segment_size{detail::round_to_pow_of_2(small_segment_size)}
+, m_large_segment_size{detail::round_to_pow_of_2(large_segment_size)}
+, m_num_huge_alloc_warn_threshold{num_huge_alloc_warn_threshold}
+{
+    // Validate that tiny_limit < small_limit < large_limit
+    if (!(m_tiny_limit < m_small_limit && m_small_limit < m_large_limit))
+    {
+        std::ostringstream os;
+        os << "Invalid heap size configuration: HWMALLOC_TINY_LIMIT < HWMALLOC_SMALL_LIMIT < HWMALLOC_LARGE_LIMIT must hold. ";
+        os << "Got HWMALLOC_TINY_LIMIT=" << m_tiny_limit
+           << ", HWMALLOC_SMALL_LIMIT=" << m_small_limit
+           << ", HWMALLOC_LARGE_LIMIT=" << m_large_limit << ".";
+        throw std::runtime_error(os.str());
+    }
+
+    // Validate that limits are at most segment sizes
+    if (!(tiny_limit <= tiny_segment_size && small_limit <= small_segment_size &&
+            large_limit <= large_segment_size))
+    {
+        std::ostringstream os;
+        os << "Invalid heap size configuration: Limits must be at most segment sizes. ";
+        os << "Got HWMALLOC_TINY_LIMIT=" << m_tiny_limit
+           << ", HWMALLOC_TINY_SEGMENT_SIZE=" << m_tiny_segment_size
+           << ", HWMALLOC_SMALL_LIMIT=" << m_small_limit
+           << ", HWMALLOC_SMALL_SEGMENT_SIZE=" << m_small_segment_size
+           << ", HWMALLOC_LARGE_LIMIT=" << m_large_limit
+           << ", HWMALLOC_LARGE_SEGMENT_SIZE=" << m_large_segment_size << ".";
+        throw std::runtime_error(os.str());
+    }
+}
+
+heap_config const&
+get_default_heap_config()
+{
+    static heap_config config{
+        detail::get_env_bool("HWMALLOC_NEVER_FREE", false),
+        detail::get_env_size_t("HWMALLOC_NUM_RESERVE_SEGMENTS", 16u),
+        detail::get_env_size_t("HWMALLOC_TINY_LIMIT", (1u << 7)),          // 128B
+        detail::get_env_size_t("HWMALLOC_SMALL_LIMIT", (1u << 12)),        // 4KiB
+        detail::get_env_size_t("HWMALLOC_LARGE_LIMIT", (1u << 21)),        // 2MiB
+        detail::get_env_size_t("HWMALLOC_TINY_SEGMENT_SIZE", (1u << 16)),  // 64KiB
+        detail::get_env_size_t("HWMALLOC_SMALL_SEGMENT_SIZE", (1u << 16)), // 64KiB
+        detail::get_env_size_t("HWMALLOC_LARGE_SEGMENT_SIZE", (1u << 21)),  // 2MiB
+        detail::get_env_size_t("HWMALLOC_NUM_HUGE_ALLOC_WARN_THERESHOLD", 10)
```

Comment:
Separate comment for default value: Probably it's good to warn immediately when going over num_reserve_segments? I.e. set the threshold to 1 or 0?